src/model/ZoneHVACEvaporativeCoolerUnit.cpp

```diff
@@ -0,0 +1,530 @@
+*  ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+***********************************************************************************************************************/
+
+#include "ZoneHVACEvaporativeCoolerUnit.hpp"
+#include "ZoneHVACEvaporativeCoolerUnit_Impl.hpp"
+
+#include "Model.hpp"
+#include "Model_Impl.hpp"
+#include "Schedule.hpp"
+#include "Schedule_Impl.hpp"
+#include "HVACComponent.hpp"
+#include "HVACComponent_Impl.hpp"
+#include "ScheduleTypeLimits.hpp"
+#include "ScheduleTypeRegistry.hpp"
+#include "FanComponentModel.hpp"
+#include "FanComponentModel_Impl.hpp"
+#include "EvaporativeCoolerDirectResearchSpecial.hpp"
+#include "EvaporativeCoolerDirectResearchSpecial_Impl.hpp"
+
+#include "../utilities/core/Assert.hpp"
+#include "../utilities/data/DataEnums.hpp"
+
+#include <utilities/idd/IddFactory.hxx>
+#include <utilities/idd/IddEnums.hxx>
+#include <utilities/idd/OS_ZoneHVAC_EvaporativeCoolerUnit_FieldEnums.hxx>
+
+namespace openstudio {
+namespace model {
+
+  namespace detail {
+
+    ZoneHVACEvaporativeCoolerUnit_Impl::ZoneHVACEvaporativeCoolerUnit_Impl(const IdfObject& idfObject, Model_Impl* model, bool keepHandle)
+      : ZoneHVACComponent_Impl(idfObject, model, keepHandle) {
+      OS_ASSERT(idfObject.iddObject().type() == ZoneHVACEvaporativeCoolerUnit::iddObjectType());
+    }
+
+    ZoneHVACEvaporativeCoolerUnit_Impl::ZoneHVACEvaporativeCoolerUnit_Impl(const openstudio::detail::WorkspaceObject_Impl& other, Model_Impl* model,
+                                                                           bool keepHandle)
+      : ZoneHVACComponent_Impl(other, model, keepHandle) {
+      OS_ASSERT(other.iddObject().type() == ZoneHVACEvaporativeCoolerUnit::iddObjectType());
+    }
+
+    ZoneHVACEvaporativeCoolerUnit_Impl::ZoneHVACEvaporativeCoolerUnit_Impl(const ZoneHVACEvaporativeCoolerUnit_Impl& other, Model_Impl* model,
+                                                                           bool keepHandle)
+      : ZoneHVACComponent_Impl(other, model, keepHandle) {}
+
+    ModelObject ZoneHVACEvaporativeCoolerUnit_Impl::clone(Model model) const {
+      auto evaporativeCoolUnitClone = ZoneHVACComponent_Impl::clone(model).cast<ZoneHVACEvaporativeCoolerUnit>();
+
+      if (OptionalHVACComponent intermediate = optionalSupplyAirFan()) {
+        evaporativeCoolUnitClone.setSupplyAirFan(intermediate->clone(model).cast<HVACComponent>());
+      }
+      if (OptionalHVACComponent intermediate = optionalFirstEvaporativeCooler()) {
+        evaporativeCoolUnitClone.setFirstEvaporativeCooler(intermediate->clone(model).cast<HVACComponent>());
+      }
+      if (OptionalHVACComponent intermediate = secondEvaporativeCooler()) {
+        evaporativeCoolUnitClone.setSecondEvaporativeCooler(intermediate->clone(model).cast<HVACComponent>());
+      }
+
+      return std::move(evaporativeCoolUnitClone);
+    }
+
+    std::vector<IdfObject> ZoneHVACEvaporativeCoolerUnit_Impl::remove() {
```

Comment:
Could consider using `#include "../utilities/core/ContainersMove.hpp"`, eg https://github.com/NREL/OpenStudio/blob/0314c3ea3e96691688ccb138b442684347f5fdde/src/model/Building.cpp#L122

Reply:
Also, given the way you wrote the `children()` method, I'm not sure explicitly cloning the children is necessary, but I could be wrong. In the end, I only care if the (future) test shows everything works as expected.